src/main/java/com/google/devtools/build/lib/bazel/rules/cpp/BazelCppSemantics.java

```diff
@@ -42,12 +42,16 @@ public PathFragment getEffectiveSourcePath(Artifact source) {
   @Override
   public void finalizeCompileActionBuilder(
       RuleContext ruleContext, CppCompileActionBuilder actionBuilder) {
-    actionBuilder.setCppConfiguration(ruleContext.getFragment(CppConfiguration.class));
+    CppConfiguration cppConfiguration = ruleContext.getFragment(CppConfiguration.class);
+    actionBuilder.setCppConfiguration(cppConfiguration);
     actionBuilder.setActionContext(CppCompileActionContext.class);
-    // Because Bazel does not support include scanning, we need the entire crosstool filegroup,
-    // including header files, as opposed to just the "compile" filegroup.
-    actionBuilder.addTransitiveMandatoryInputs(CppHelper.getToolchain(ruleContext).getCrosstool());
-    actionBuilder.setShouldScanIncludes(false);
+    boolean shouldScanIncludes = cppConfiguration.shouldScanIncludes();
+    if (!shouldScanIncludes) {
+      // When not performing include scanning, we need the entire crosstool filegroup,
+      // including header files, as opposed to just the "compile" filegroup.
+      actionBuilder.addTransitiveMandatoryInputs(CppHelper.getToolchain(ruleContext).getCrosstool());
+    }
+    actionBuilder.setShouldScanIncludes(shouldScanIncludes);
```

Comment:
`genrule` infrastructure? What's that?

include scanning is custom C++ specific logic in Blaze. It serves two purposes: add includes that are not specified by the rule but are still needed (bad) and remove includes from the inputs of the action that are not actually needed (good, helps with action caching especially when remote executors are in play)